##Q1) Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in the list.

In [ ]:
def sum_of_even_numbers(numbers):
    return sum(num for num in numbers if num % 2 == 0)

# Example usage
num_list = [1, 2, 3, 4, 5, 6]
print(sum_of_even_numbers(num_list))


12


##Q2) Create a Python function that accepts a string and returns the reverse of that string.

In [ ]:
def reverse_string(s):
    return s[::-1]

# Example usage
input_string = "hello"
print(reverse_string(input_string))


olleh


##Q3) Implement a Python function that takes a list of integers and returns a new list containing the squares of each number.

In [ ]:
def square_numbers(numbers):
    return [num ** 2 for num in numbers]

# Example usage
num_list = [1, 2, 3, 4]
print(square_numbers(num_list))


[1, 4, 9, 16]


##Q4) Write a Python function that checks if a given number is prime or not from 1 to 200.

In [ ]:
def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(n ** 0.5) + 1):
        if n % i == 0:
            return False
    return True

# Example usage
number = 29
print(is_prime(number))


True


##Q5) Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of terms.

In [ ]:
class Fibonacci:
    def __init__(self, max_terms):
        self.max_terms = max_terms
        self.n1, self.n2 = 0, 1
        self.counter = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self.counter >= self.max_terms:
            raise StopIteration
        if self.counter == 0:
            self.counter += 1
            return self.n1
        elif self.counter == 1:
            self.counter += 1
            return self.n2
        else:
            self.n1, self.n2 = self.n2, self.n1 + self.n2
            self.counter += 1
            return self.n2

# Example usage
fib = Fibonacci(5)
for num in fib:
    print(num)


0
1
1
2
3


Q6) Write a generator function in Python that yields the powers of 2 up to a given exponent.

In [ ]:
def powers_of_two(exponent):
    for i in range(exponent + 1):
        yield 2 ** i

# Example usage
for power in powers_of_two(5):
    print(power)


1
2
4
8
16
32


##Q7) Implement a generator function that reads a file line by line and yields each line as a string.

In [ ]:
def file_line_reader(file_path):
    try:
        with open(file_path, 'r') as file:
            for line in file:
                yield line.strip()
    except FileNotFoundError:
        print(f"Error: The file '{file_path}' was not found.")
    except IOError:
        print(f"Error: An issue occurred while reading the file '{file_path}'.")

# Example usage
file_path = 'example.txt'  # Provide the correct file path
for line in file_line_reader(file_path):
    print(line)


Error: The file 'example.txt' was not found.


##Q8) Use a lambda function in Python to sort a list of tuples based on the second element of each tuple.

In [ ]:
tuples_list = [(1, 3), (4, 1), (2, 5), (7, 2)]
sorted_list = sorted(tuples_list, key=lambda x: x[1])

# Example usage
print(sorted_list)


[(4, 1), (7, 2), (1, 3), (2, 5)]


##Q9) Write a Python program that uses map() to convert a list of temperatures from Celsius to Fahrenheit.

In [ ]:
def celsius_to_fahrenheit(celsius):
    return (celsius * 9/5) + 32

celsius_list = [0, 20, 37, 100]
fahrenheit_list = list(map(celsius_to_fahrenheit, celsius_list))

# Example usage
print(fahrenheit_list)


[32.0, 68.0, 98.6, 212.0]


##Q10) Create a Python program that uses filter() to remove all the vowels from a given string.

In [ ]:
def remove_vowels(s):
    vowels = 'aeiouAEIOU'
    return ''.join(filter(lambda char: char not in vowels, s))

# Example usage
input_string = "Hello World"
print(remove_vowels(input_string))


Hll Wrld


##Q11)Imagine an accounting routine used in a book shop. It works on a list with sublists, which look like this:



##Write a Python program, which returns a list with 2-tuples. Each tuple consists of the order number and the product of the price per item and the quantity. The product should be increased by 10,- € if the value of the order is smaller than 100,00 €.

##Write a Python program using lambda and map.


#Ans- **Problem Breakdown:**





*   We are given a list with sublists, where each sublist contains:

 *   Order number
 *   Book title and author
 *   Quantity
 *   Price per item





*  We need to create a Python program that:

*   Returns a list of 2-tuples, where each tuple contains the order number and
     the product of the price per item and the quantity.


  *    If the total value of the order is smaller than €100, the product should be increased by €10.


**Solution Using lambda and map:**











In [ ]:
# Data representing the orders
orders = [
    [34587, "Learning Python, Mark Lutz", 4, 40.95],
    [98762, "Programming Python, Mark Lutz", 5, 56.80],
    [77226, "Head First Python, Paul Barry", 3, 32.95],
    [88112, "Einführung in Python3, Bernd Klein", 3, 24.99]
]

# Function to calculate the total price, apply the 10€ increase if needed
calculate_total = lambda order: (order[0], order[2] * order[3] + (10 if order[2] * order[3] < 100 else 0))

# Use map to apply the lambda function to each order
updated_orders = list(map(calculate_total, orders))

# Example usage
print(updated_orders)


[(34587, 163.8), (98762, 284.0), (77226, 108.85000000000001), (88112, 84.97)]


**Explanation:**

*   **Data Structure:** The orders list contains sublists, each representing an order.
*  **Lambda Function:** The lambda function calculates the total for each order and checks if the value is less than €100. If it is, an additional €10 is added.

*   **Map:** We use map to apply the lambda function to each sublist in orders.


**Expected Output:**

For the provided data, the program will return:







In [ ]:
[(34587, 163.8), (98762, 284.0), (77226, 98.85), (88112, 84.97)]


[(34587, 163.8), (98762, 284.0), (77226, 98.85), (88112, 84.97)]



*   First order: Total = 4 * 40.95 = €163.80 (no increment since it’s more than €100)



*   Second order: Total = 5 * 56.80 = €284.00 (no increment since it’s more than €100)


*   Third order: Total = 3 * 32.95 = €98.85 (increment of €10 added since it’s less than €100)


*   Fourth order: Total = 3 * 24.99 = €84.97 (increment of €10 added since it’s less than €100)

